src/tree_cmd.cpp

```diff
@@ -805,6 +859,110 @@ CommandCost CmdPlantTree(DoCommandFlags flags, TileIndex end_tile, TileIndex sta
 	}
 }
 
+
+int _tree_placer_cmd_count = 0;
+/**
+ * Sync trees, sent when a client is using the Tree Placer.
+ * @param flags type of operation
+ * @param cmd_data stuff
+ * @return the cost of this operation or an error
+ */
+CommandCost CmdSyncTrees(DoCommandFlags flags, const TreeSyncCmdData &cmd_data)
+{
+	if (_networking && cmd_data.calling_client == _network_own_client_id && flags.Test(DoCommandFlag::Execute)) {
+		_tree_placer_cmd_count--;
+	}
+	StringID msg = INVALID_STRING_ID;
+	CommandCost cost(EXPENSES_OTHER);
+
+	Company *c = (_game_mode != GM_EDITOR) ? Company::GetIfValid(_current_company) : nullptr;
+	int limit = (c == nullptr ? INT32_MAX : GB(c->tree_limit, 16, 16));
+
+	/* Iterate through sync_data, plant trees on every tile inside. */
+	for (std::pair<TileIndex, TreePlacerData> t : cmd_data.sync_data) {
+		if (t.first >= Map::Size()) return CMD_ERROR;
+
+		uint8_t tree_count = (IsTileType(t.first, MP_TREES)) ? t.second.count - GetTreeCount(t.first) : t.second.count;
+
+		CommandCost attempt = CmdPlantTree(flags, t.first, t.first, t.second.type, tree_count, false);
+		cost.AddCost(attempt.GetCost());
+
+		/* If we're the client that sent this command, this cleans up everything we wanted to place. */
+		if (cmd_data.calling_client == _network_own_client_id && _tree_placer_memory.contains(t.first)) {
+			_tree_placer_memory.erase(t.first);
+		}
+		
+		/* Tree limit used up? No need to check more. */
+		if (limit < 0) break;
+	}
+
+	/* Test tree limit. */
+	if (--limit < 1) {
+		msg = STR_ERROR_TREE_PLANT_LIMIT_REACHED;
+	}
+
+	if (cost.GetCost() == 0) {
+		return CommandCost(msg);
+	} else {
+		return cost;
+	}
+}
+
+void TreeSyncCmdData::Serialise(BufferSerialisationRef buffer) const
+{
+	buffer.Send_uint32(this->calling_client);
+	buffer.Send_uint32((uint32_t)this->sync_data.size());
+	for (std::pair<TileIndex, TreePlacerData> t : this->sync_data) {
+		buffer.Send_uint32(t.first.value);
+		buffer.Send_uint8(t.second.type);
+		buffer.Send_uint8(t.second.count);
+	}
+}
+
+bool TreeSyncCmdData::Deserialise(DeserialisationBuffer &buffer, StringValidationSettings default_string_validation)
+{
+	this->calling_client = (ClientID)buffer.Recv_uint32();
+	uint32_t size = buffer.Recv_uint32();
+	if (!buffer.CanRecvBytes(size * 4)) return false;
+	for (; size > 0; size--) {
+		TileIndex tile = TileIndex{buffer.Recv_uint32()};
+		TreeType type = (TreeType)buffer.Recv_uint8();
+		uint8_t count = buffer.Recv_uint8();
+		this->sync_data[tile] = TreePlacerData{type, count};
+	}
+	return true;
+}
+
+void TreeSyncCmdData::FormatDebugSummary(format_target &output) const
+{
+	output.format("Size {}", this->sync_data.size());
+}
+
+static int last_cmd_limit_tick_count = 0;
+void SendSyncTrees()
+{
+	if (_tree_placer_memory.size() == 0) return;
+	if (_networking) {
+		if (_tree_placer_cmd_count > 4) return;
+		_tree_placer_cmd_count++;
```

Comment:
If the DoCommandP fails for whatever reason this might end up stuck at a positive value even with nothing in flight, but this might not any real problem in practice.